### PR TITLE
Making `iree-util-strip-and-splat-constants` use byte_patterns.

### PR DESCRIPTION
**Pull request by benvanik**
Had to remove the path someone added for arith.constant as that can't take byte patterns (and it wasn't the intent of the pass). If someone wants that they can hoist into globals first before running this pass with `--iree-util-outline-constants`.

A common usage of this would be to take a giant mlir file and run it through iree-opt with
`--iree-util-outline-constants --iree-util-strip-and-splat-constants` to outline any arith.constant values into globals and then strip them.

Note that the final VMFB will contain the same number of bytes as the original unstripped file.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Util/Transforms/StripAndSplatConstants.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    auto moduleOp = getOperation();
+    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
+
+    // Use a heuristic to space out splat values so it's clearer where each in
+    // the final file comes from.
+    int replaceIndex = 1;
+    auto getSplatAttr = [&](ShapedType type) {
+      return IREE::Util::BytePatternAttr::get(moduleOp.getContext(), type,
+                                              replaceIndex++);
+    };
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/StripAndSplatConstants.cpp]: We had a heuristic in here before:

```
    // Use a heuristic to space out splat values in hopes of avoiding NaN and
    // INF values at runtime:
    //   floats: 1/1, 1/2, 1/3, ...
    //   ints: 1, 2, 3, 4, ...
```
```c++
      if (elementType.isa<FloatType>()) {
        newValue = DenseElementsAttr::get(
            tensorType, FloatAttr::get(elementType, 1.0 / replaceIndex));
      } else {
        newValue = DenseElementsAttr::get(
            tensorType, IntegerAttr::get(elementType, replaceIndex));
      }
```

This is all for easier development/debugging, so fine to just iterate based on what is observed in workloads.

IIRC, the thinking at the time was that
* if all ML model weights were positive integer values, they would eventually overflow after multiplications
* if a single value like 1 was chosen, code like `x / (x - x)` would divide by zero

File: compiler/src/iree/compiler/Dialect/Util/Transforms/StripAndSplatConstants.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+public:
+  StripAndSplatConstantsPass() = default;
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Util::UtilDialect>();
+  }
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/StripAndSplatConstants.cpp]: https://github.com/openxla/iree/actions/runs/5825552986/job/15797405535?pr=14627#step:4:5992
```
/work/compiler/src/iree/compiler/Dialect/Util/Transforms/StripAndSplatConstants.cpp:35:10: error: unused variable 'builder' [-Werror,-Wunused-variable]
    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
         ^
```

**Review by ScottTodd — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/Util/Transforms/StripAndSplatConstants.cpp (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/StripAndSplatConstants.cpp]: yeah, now with byte_pattern nothing can fold with these values (so no issues in the compiler) and we don't really care about execution results when all the constants are stripped (as performance and correctness are both invalid to inspect with stripped constants)

**Comment by stellaraccident** [comment]
Thank you!

**Pull request merged**